### PR TITLE
Drop cameras from the observatory if the coarse focus fails.

### PR DESCRIPTION
**Pull request by wtgee**
This just checks which of the cameras successfully set the autofocus event and removes the ones that did not.

Related to #304

### CODE REVIEW

**Comment by wtgee** [comment]
I should write a test...

File: src/huntsman/pocs/states/huntsman/coarse_focusing.py [hunk 1]
```diff
@@ -1,7 +1,6 @@
-from panoptes.utils.time import wait_for_events, current_time
-
-from panoptes.utils import altaz_to_radec
-from panoptes.pocs.utils.location import create_location_from_config
+from panoptes.utils.time import CountdownTimer
```
Comment (JAAlvarado-Montes) [inline src/huntsman/pocs/states/huntsman/coarse_focusing.py:1]:
```suggestion

```

File: src/huntsman/pocs/states/huntsman/coarse_focusing.py [hunk 2]
```diff
@@ -1,7 +1,6 @@
...
+from panoptes.utils.time import current_time
```
Comment (JAAlvarado-Montes) [inline src/huntsman/pocs/states/huntsman/coarse_focusing.py:2]:
```suggestion
from panoptes.utils.time import CountdownTimer, current_time
```

**Review by JAAlvarado-Montes — APPROVED** [review]
LGTM! Just a small suggestion, or maybe there's a reason I don't know for having the import like that :)

File: src/huntsman/pocs/states/huntsman/coarse_focusing.py (hunk 2 above)
Comment (wtgee) [reply src/huntsman/pocs/states/huntsman/coarse_focusing.py:2]: Unfortunately we don't have a set style policy...

File: src/huntsman/pocs/states/huntsman/coarse_focusing.py [hunk 3]
```diff
@@ -27,24 +24,48 @@ def on_enter(event_data):
...
+        # If overall timer has expired, check which cameras have not finished
+        # and remove them from the observatory.
+        if timer.expired():
+            pocs.logger.info(f'Autofocus timer expired, checking which cameras were successful')
+            for cam_name, focus_event in autofocus_events.items():
+                if not focus_event.is_set():
+                    pocs.logger.warning(f'Incomplete focus event on {cam_name}, removing')
+                    # TODO figure out why and also mark them for restart somehow.
+                    # See https://github.com/AstroHuntsman/huntsman-pocs/issues/304
+                    pocs.observatory.remove_camera(cam_name)
```
Comment (danjampro) [inline src/huntsman/pocs/states/huntsman/coarse_focusing.py:58]: I've seen problems before where changing the size of dictionaries while iterating over them can cause errors (e.g., https://stackoverflow.com/questions/11941817/how-to-avoid-runtimeerror-dictionary-changed-size-during-iteration-error).

I'm not sure whether we will have that problem here? If this is already tested in `POCS` then it should be fine.

File: src/huntsman/pocs/states/huntsman/coarse_focusing.py [hunk 4]
```diff
@@ -27,24 +24,48 @@ def on_enter(event_data):
...
-    wait_for_events(list(autofocus_events.values()), timeout=coarse_focus_timeout)
 
-    # Morning and not dark enough for observing...
-    if pocs.observatory.past_midnight and not pocs.is_dark(horizon='observe'):
-        pocs.next_state = 'twilight_flat_fielding'
-    else:
-        pocs.next_state = 'scheduling'
+    while True:
+        # Check if all cameras are finished and proceed if so.
+        if all([e.is_set() for e in autofocus_events.values()]):
```
Comment (danjampro) [inline src/huntsman/pocs/states/huntsman/coarse_focusing.py:46]: We are still using events for pyro autofocus, but at some stage we should switch to `concurrent.futures`, which implements a nice `wait` function (https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.wait) that will all us to use the timeout intrinsic to the focuser, which is a bit cleaner.

**Review by danjampro — APPROVED** [review]
Hi @wtgee looks good to me, provided that changing the dictionary size in the loop doesn't raise an error (see comment).

File: src/huntsman/pocs/states/huntsman/coarse_focusing.py (hunk 3 above)
Comment (wtgee) [reply src/huntsman/pocs/states/huntsman/coarse_focusing.py:58]: This is iterating over a a dictionary of just camera names and `threading.Event` objects, so I think we should be okay.

**Pull request merged**